walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,207 @@
+                frequency.put(slot, prevOptionalAttendee);
+            }
+        }
+      }
+      
+      // Return a collection of times with the maximum number of optional attendees
+      return getKeysWithMaximumValue(frequency);
+  }
+
+  public Collection<TimeRange> getTimeSlotWithMandatoryAttendeesOnly(ArrayList<TimeRange> times, MeetingRequest request, int startTime, int endTime){
+      
+    ArrayList<TimeRange> openTime = new ArrayList<>();
+
+    // Edge case if there are no conflicting events
+    if (times.isEmpty()) {
+        return Arrays.asList(TimeRange.WHOLE_DAY);
+    }
+
+    // Sort all event by ascending start times O(nlogn)
+    times.sort(TimeRange.ORDER_BY_START);
+
+    // Create dummy "start of day" & "end of day" events
+    times.add(0, TimeRange.fromStartEnd(startTime, startTime, false));
+    times.add(TimeRange.fromStartEnd(endTime, endTime, false));
+
+    // Maintain a pointer for optimal ending points
+    TimeRange prevEvent = times.get(0);
+    int optimalEndTime = prevEvent.end();
+
+    for (int i = 1; i < times.size(); i++){
+        TimeRange nextEvent = times.get(i);
+
+        // Move on from overlapping events
+        if (prevEvent.overlaps(nextEvent)) {
+            optimalEndTime = Math.max(prevEvent.end(), nextEvent.end());
+            prevEvent = nextEvent;
+            continue;
+        }
+
+        // Non-overlapping events
+        int duration = nextEvent.start() - optimalEndTime;
+        if (duration >= request.getDuration()){
+            // Deals with inclusivity if this is the last event
+            if (i == times.size() - 1) {
+                openTime.add(TimeRange.fromStartEnd(optimalEndTime, nextEvent.start(), true));
+            } else {
+                openTime.add(TimeRange.fromStartEnd(optimalEndTime, nextEvent.start(), false));
+            }
+        }
+
+        optimalEndTime = nextEvent.end();
+        prevEvent = nextEvent;
+    }
+
+    return openTime;
+  }
+
+  public Collection<TimeRange> fullDayAsOpenSlot(){
+      return Arrays.asList(TimeRange.WHOLE_DAY);
+  }
+
+  public Collection<TimeRange> getKeysWithMaximumValue(Map<TimeRange, Integer> frequency){
+
+    // Return the time slot with the max # of mandatory & optional attendees
+    Map.Entry<TimeRange, Integer> maxEntry = null;
+    for (Map.Entry<TimeRange, Integer> entry : frequency.entrySet()) {
+        if (maxEntry == null || entry.getValue() > maxEntry.getValue()) {
+            maxEntry = entry;
+        }
+    }
+    Integer maxAttendees = maxEntry.getValue();
+
+    // Run through frequency table to collect timeslots with the maximum attendees
+    Collection<TimeRange> timeSlots = new ArrayList<>();
+    for (Map.Entry<TimeRange, Integer> entry : frequency.entrySet()) {
+        if (entry.getValue() ==  maxAttendees) {
+            timeSlots.add(entry.getKey());
+        }
+    }
+    return timeSlots;
   }
```

Comment:
This function looks good but a small nitpick is that it could be done with one for loop where you clear the list if you find a new maximum.